src/compiler/checker.ts

```diff
@@ -5230,15 +5234,20 @@ namespace ts {
             return links.resolvedType;
         }
 
-        function createTupleType(elementTypes: Type[]) {
-            const id = getTypeListId(elementTypes);
-            return tupleTypes[id] || (tupleTypes[id] = createNewTupleType(elementTypes));
+        function createTupleType(elementTypes: Type[], thisType?: Type) {
+            let id = getTypeListId(elementTypes);
+            if (thisType) {

```

Comment:
You need to always append an extra id no matter whether you have a this-type or not. Otherwise you end up confusing tuples with a this-type with next-higher-arity tuples without a this-type. I suggest always using

```
let id = getTypeListId(elementTypes) + "," + thisType ? thisType.id : 0;
```

Then also change `createType` to increment `typeCount` before assigning it to the new type's `id` property.